plugins/widget/plugin.js

```diff
@@ -3310,70 +3327,268 @@
+	}
+
+	// Replace parts object containing:
+	// partName => selector pairs
+	// with:
+	// partName => element pairs
+	function setupParts( widget ) {
+		if ( widget.parts ) {
+			var parts = {},
+				el,
+				partName;
+
+			for ( partName in widget.parts ) {
+				el = widget.wrapper.findOne( widget.parts[ partName ] );
+				parts[ partName ] = el;
+			}
+			widget.parts = parts;
 		}
+	}
 
-		// Setting a listener on data is enough, there's no need to perform it on widget initialization, as
-		// setupWidgetData fires this event anyway.
-		widget.on( 'data', function() {
-			// In some cases widget might get destroyed in an earlier data listener. For instance, image2 plugin, does
-			// so when changing its internal state.
-			if ( !widget.wrapper ) {
-				return;
+	function setupEditables( widget ) {
+		var definedEditables = widget.editables,
+			editableName,
+			editableDef;
+
+		widget.editables = {};
+
+		if ( !widget.editables ) {
+			return;
+		}
+
+		for ( editableName in definedEditables ) {
+			editableDef = definedEditables[ editableName ];
+			widget.initEditable( editableName, typeof editableDef == 'string' ? { selector: editableDef } : editableDef );
+		}
+	}
+
+	function setupMask( widget ) {
+		if ( widget.mask === true ) {
+			setupFullMask( widget );
+		} else if ( widget.mask ) {
+			// Buffer to limit number of separate calls to 'setupPartialMask', e.g. during writing.
+			var maskBuffer = new CKEDITOR.tools.buffers.throttle( 250, setupPartialMask, widget ),
+				changeListener,
+				blurListener;
+
+			widget.on( 'focus', function() {
+				changeListener = widget.editor.on( 'change', maskBuffer.input );
+				blurListener = widget.on( 'blur', function() {
+					changeListener.removeListener();
+					blurListener.removeListener();
+				} );
+			} );
+
+			// Focusing editable doesn't trigger focus on widget, so listen to those events separately.
+			for ( var editable in widget.editables ) {
+				widget.editables[ editable ].on( 'focus', function() {
+					widget.editor.on( 'change', maskBuffer.input );
+					// If widget was focused before focusing editable, the 'blur' event has to be removed.
+					// Otherwise on Chrome it will trigger after the focus event and cancel listening to
+					// changes (on FF it works inversely).
+					if ( blurListener ) {
+						blurListener.removeListener();
+					}
+				} );
+				widget.editables[ editable ].on( 'blur', function() {
+					widget.editor.removeListener( 'change', maskBuffer.input );
+				} );
 			}
+			maskBuffer.input();
 
-			var label = this.getLabel ? this.getLabel() : getLabelDefault.call( this );
+			// FF renders image-like widget very late (much after 'instanceReady') so mask creation
+			// has to be postponed. The same is true for switching between WYSIWYG and source mode,
+			// just different event and smaller delay are needed.
+			if ( CKEDITOR.env.gecko ) {
+				widget.editor.on( 'instanceReady', function() {
```

Comment:
I know we have already talked about it, but timeouts are quite a last resort solutions (and could be quite unreliable). We also talked about `img.complete` property, that it doesn't do the job, but have you tried to combine it with `img.load` event like below:

```js
var imgs = widget.element.find( 'img' );
CKEDITOR.tools.array.forEach( imgs.toArray(), function( img ) {
	if ( !img.$.complete ) {
		img.on( 'load', function() {
			maskBuffer.input();
		} );
	}
} );
```

It seems to be working covering this behaviour in FF (I have also tested it with services like https://picsum.photos which loads different photos each time so it's not cached and with throttling in dev console and it works fine). Could you check this one? Of if you already checked it, described a case in which it failed?

Reply:
I just tested the code you proposed. The problem with it occurs when the images aren't different each time, to be more precisely - when image is already cached. Then it seems to always have `img.complete === true` and doesn't attach listener on `load` event.
So I tried to abandon this condition and just always attach listener (which is obviously less efficient) and it works pretty ok, in fact I still managed to break it a few times but `1.` it was rather rare (about 1 per 20-30 page refreshes) and `2.` I can't guarantee that the reason wasn't connected to abusive browser refreshing. So I think it can be called quite reliable solution.